enhancements/operator-dependency-resolution.md

```diff
@@ -0,0 +1,222 @@
+  displayName: CatalogSource Test
+  sourceType: grpc
+  image: docker.io/test/catsrc-test:latest
+  indexNamespace: io.redhat
+```
+
+An example of a `dependencies.yaml` that specifies Prometheus operator and etcd CRD dependencies:
+
+```
+dependencies:
+  - type: olm.package
+    indexNamespace: io.redhat
+    name: prometheus
+    version: >0.27.0
+  - type: olm.gvk
+    name: etcdclusters.etcd.database.coreos.com
+    group: etcd.database.coreos.com
+    kind: EtcdCluster
+    version: v1beta2
+```
+
+#### Provided APIs
+
+The two default types of provided APIs (operator name + version and CRD GVK) will be inferred just by the registry parsing the manifests provided in the bundle. However, we acknowledge that in the future there may be additional APIs provided by operators that are not as easily discoverable.
+
+There are cases when the operators provide/own certain APIs that are important to consider during dependency resolution in order to allow the operator to work properly. This information needs to be declared in the bundle so it can be queried later. The proposed syntax for provided APIs in `dependencies.yaml`:
+```
+provided:
+  - type: istio
+    version: >1.14.0
+```
+
+That being said, these additional types will need to be implemented as they become more clear. For now, `dependencies.yaml` will not include any explicit reference to provided APIs.
+
+### Cluster Admin UX
+
+Given the changes to the way OLM will resolve dependencies, it is also necessary to push more detailed data to cluster administrators about what dependencies were installed (or will be installed) and why. To that end, we will continue to use the InstallPlan status to push detailed resolution information to the user. This will include data about what dependencies were installed along with information about where child dependencies came from (essentially, the result of the resolver in detail).
+
+There was also some initial discussion around what the user workflow would be to edit/override dependencies. Given the current state of the APIs that exist (specifically the Subscription and InstallPlan APIs), those would require extensive changes to create a trivial override for dependencies. Even then, there is a configurability problem where a single install overriding a dependency may not enough and a cluster admin may want to create a cluster wide or namespaced override that is persistent and editable. Initially, there was some discussion about the possibility of a dependency Override API or the possibility of using the Operator API for this purpose, but those discussions are ultimately outside the scope of the immediate (3+ month) timeframe and scope of this enhancement proposal, and will be taken up further in the future.
+
+### User Stories
+
+#### Explicit Operator Dependency Metadata
+
+As an operator author, I want to specify dependencies explicitly in the bundle. For example, I would like to specify that my operator depends on etcd operator version 0.9.0.
```

Comment:
What happens if two catalogs exist that provide different operators with the same name/version? I assume the `indexNamespace` API is meant to solve this issue?

Reply:
Correct. `indexNamespace` is a way for OLM to resolve dependencies in deterministic way. This may be 4.6+ feature. For 4.5, we plan to do a CatalogSource weighting which essentially adds the order to CatalogSource so there is always a priority choice.